src/operator/output.rs

```diff
@@ -0,0 +1,311 @@
+        }
+
+        Self { mailbox }
+    }
+
+    fn take_from_worker(&self, worker: usize) -> Option<T> {
+        self.mailbox[worker].take()
+    }
+
+    fn mailbox(&self, worker: usize) -> &Mailbox<Option<T>> {
+        &self.mailbox[worker]
+    }
+}
+
+/// A handle used to read data from a stream from outside the circuit.
+///
+/// Internally, the handle manages an array of mailboxes, one for
+/// each worker thread.  At the end of each clock cycle, the worker
+/// writes the current contents of the stream to the mailbox.
+///
+/// Once the clock cycle ends (i.e., the
+/// [`DBSPHandle::step`](`crate::DBSPHandle::step`) method
+/// returns), each mailbox contains a single value -- a copy of
+/// stream contents at the current clock cycle.
+///
+/// The client retrieves values produced by individual workers using
+/// the [`take_from_worker`](`OutputHandle::take_from_worker`) method.
+/// Alternatively they can retrieve values from all mailboxes at once
+/// using [`take_from_all`](`OutputHandle::take_from_all`).
+/// If the stream carries relational data, the
+/// [`consolidate`](`OutputHandle::consolidate`) method can be used
+/// to combine output batches produced by all workes into a single
+/// batch.
+///
+/// Reading from a mailbox using any of these methods removes the value
+/// leaving the mailbox empty.  If the value is not read, it gets
+/// overwritten at the next clock cycle (i.e., during the next call to
+/// `step`).
+#[derive(Clone)]
+pub struct OutputHandle<T>(Arc<OutputHandleInternal<T>>);
+
+impl<T> OutputHandle<T>
+where
+    T: Send + Clone + 'static,
+{
+    fn new() -> Self {
+        match Runtime::runtime() {
+            None => Self(Arc::new(OutputHandleInternal::new(1))),
+            Some(runtime) => {
+                let output_id = runtime.sequence_next(Runtime::worker_index());
+
+                runtime
+                    .local_store()
+                    .entry(OutputId::new(output_id))
+                    .or_insert_with(|| {
+                        Self(Arc::new(OutputHandleInternal::new(runtime.num_workers())))
+                    })
+                    .value()
+                    .clone()
+            }
+        }
+    }
+
+    fn mailbox(&self, worker: usize) -> &Mailbox<Option<T>> {
+        self.0.mailbox(worker)
+    }
+
+    /// Read the value produced by `worker` worker thread during the last
+    /// clock cycle.
+    ///
+    /// This method is invoked between two subsequent
+    /// [`DBSPHandle::step`](`crate::DBSPHandle::step`)
+    /// calls to retrieve the value written to the stream during the last
+    /// clock cycle.  The first call is guaranteed to return a value
+    /// (since a synchronous circuit writes exactly one value to each
+    /// stream at every clock cycle).  It removes the value from the
+    /// mailbox, so subsequent calls will return `None`.
+    ///
+    /// Invoking this method in the middle of a clock cycle, i.e., during
```

Comment:
Shouldn't this be rather be undefined?

Reply:
Unspecified at the very most, it's safe code so it shouldn't be able to cause undefined behavior

Reply:
it's not undefined or unsafe, but the outcome depends on the relative ordering of operations.